### ISSUE TITLE
Test `fd/fmt` fails to compile

### ISSUE THREAD

**Issue by Kostr** [openbmc/stdplus#6]
Currently local build fails with the following error:
```
[56/138] Compiling C++ object test/fd_fmt.p/fd_fmt.cpp.o
FAILED: test/fd_fmt.p/fd_fmt.cpp.o
c++ -Itest/fd_fmt.p -Iinclude -I../include -Iinclude-fd -I../include-fd -I../subprojects/fmt/include -Isubprojects/fmt/__CMake_build -I../subprojects/fmt/__CMake_build -Isubprojects/fmt -I../subprojects/fmt -I../subprojects/function2/include -Isubprojects/function2/__CMake_build -I../subprojects/function2/__CMake_build -Isubprojects/function2 -I../subprojects/function2 -I../subprojects/googletest/googletest/include -I../subprojects/googletest/googletest -Isubprojects/googletest/__CMake_build -I../subprojects/googletest/__CMake_build -Isubprojects/googletest -I../subprojects/googletest -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c++23 -O0 -g -pthread -isystem../subprojects/googletest/googletest -isystem../subprojects/googletest/googletest/include -Wno-missing-braces -MD -MQ test/fd_fmt.p/fd_fmt.cpp.o -MF test/fd_fmt.p/fd_fmt.cpp.o.d -o test/fd_fmt.p/fd_fmt.cpp.o -c ../test/fd/fmt.cpp
In file included from ../subprojects/fmt/include/fmt/format.h:54,
                 from ../subprojects/fmt/include/fmt/compile.h:13,
                 from ../test/fd/fmt.cpp:1:
../subprojects/fmt/include/fmt/base.h: In instantiation of ‘OutputIt fmt::v10::detail::copy(InputIt, InputIt, OutputIt) [with T = char; InputIt = const char*; OutputIt = std::back_insert_iterator<stdplus::BasicStrBuf<char, 128> >; typename std::enable_if<is_back_insert_iterator<OutputIt>::value, int>::type <anonymous> = 0]’:
../subprojects/fmt/include/fmt/compile.h:151:24:   required from ‘constexpr OutputIt fmt::v10::detail::field<Char, T, N>::format(OutputIt, const Args& ...) const [with OutputIt = std::back_insert_iterator<stdplus::BasicStrBuf<char, 128> >; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; Char = char; T = std::__cxx11::basic_string<char>; int N = 0]’
../subprojects/fmt/include/fmt/compile.h:446:19:   required from ‘constexpr OutputIt fmt::v10::format_to(OutputIt, const CompiledFormat&, const Args& ...) [with OutputIt = std::back_insert_iterator<stdplus::BasicStrBuf<char, 128> >; CompiledFormat = detail::field<char, std::__cxx11::basic_string<char>, 0>; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; typename std::enable_if<detail::is_compiled_format<CompiledFormat>::value, int>::type <anonymous> = 0]’
../subprojects/fmt/include/fmt/compile.h:486:26:   required from ‘constexpr OutputIt fmt::v10::format_to(OutputIt, const S&, Args&& ...) [with OutputIt = std::back_insert_iterator<stdplus::BasicStrBuf<char, 128> >; S = stdplus::fd::FormatBuffer_Basic_Test::TestBody()::<lambda()>::FMT_COMPILE_STRING; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; typename std::enable_if<detail::is_compiled_string<S>::value, int>::type <anonymous> = 0]’
../include-fd/stdplus/fd/fmt.hpp:47:23:   required from ‘void stdplus::fd::FormatBuffer::append(const T&, Args&& ...) [with T = stdplus::fd::FormatBuffer_Basic_Test::TestBody()::<lambda()>::FMT_COMPILE_STRING; Args = {std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >}; typename std::enable_if<fmt::v10::detail::is_compiled_string<S>::value, bool>::type <anonymous> = true]’
../test/fd/fmt.cpp:30:19:   required from here
../subprojects/fmt/include/fmt/base.h:1189:28: error: no matching function for call to ‘stdplus::BasicStrBuf<char, 128>::append(const char*&, const char*&)’
 1189 |   get_container(out).append(begin, end);
      |   ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
In file included from ../include-fd/stdplus/fd/fmt.hpp:5,
                 from ../test/fd/fmt.cpp:4:
../include/stdplus/str/buf.hpp:297:22: note: candidate: ‘constexpr CharT* stdplus::BasicStrBuf<CharT, ObjSize, Allocator>::append(std::size_t) [with CharT = char; long unsigned int ObjSize = 128; Allocator = std::allocator<char>; std::size_t = long unsigned int]’
  297 |     constexpr CharT* append(std::size_t amt)
      |                      ^~~~~~
../include/stdplus/str/buf.hpp:297:22: note:   candidate expects 1 argument, 2 provided
[85/138] Compiling C++ object test/fd_ops.p/fd_ops.cpp.o
```

Steps to reproduce:
```
meson setup build && cd build && meson compile
```

**Comment by williamspatrick** [comment]
Which version of libfmt was this tested with?

**Comment by Kostr** [comment]
`HEAD` https://github.com/openbmc/stdplus/blob/master/subprojects/fmt.wrap

**Comment by wak-google** [comment]
Are you sure you have a good version of fmtlib in the subprojects?

**Comment by Kostr** [comment]
> Are you sure you have a good version of fmtlib in the subprojects?

I use the one pointed by the `*.wrap` file in the subprojects directory, which is pointing to HEAD of the [https://github.com/fmtlib/fmt](https://github.com/fmtlib/fmt)

You can see that the log message refers to the code in master.

stdplus uses `fmt::format_to`
https://github.com/openbmc/stdplus/blob/master/include-fd/stdplus/fd/fmt.hpp

`fmt::format_to(OutputIt out, const S&, Args&&... args)` uses `fmt::format_to(OutputIt out, const CompiledFormat& cf, const Args&... args)`
https://github.com/fmtlib/fmt/blob/master/include/fmt/compile.h#L478

`fmt::format_to(OutputIt out, const CompiledFormat& cf, const Args&... args)` uses `format`
https://github.com/fmtlib/fmt/blob/master/include/fmt/compile.h#L444

`format` uses `copy`
https://github.com/fmtlib/fmt/blob/master/include/fmt/compile.h#L151

`copy` uses `.append`
https://github.com/fmtlib/fmt/blob/master/include/fmt/base.h#L1189

**Comment by Kostr** [comment]
I am really not an expert on this stuff but it looks like in this expression
```cpp
class FormatBuffer
{
    ...

    template <typename T, typename... Args,
              std::enable_if_t<fmt::detail::is_compiled_string<T>::value,
                               bool> = true>
    inline void append(const T& t, Args&&... args)
    {
        fmt::format_to(std::back_inserter(buf), t, std::forward<Args>(args)...);
        writeIfNeeded();
    }

    ...
}
```
the type is deducted from the  `std::back_inserter(buf)`. In this case `buf` has a type `stdplus::StrBuf`

```cpp
class FormatBuffer
{
    ...

  private:
    stdplus::StrBuf buf;
    ...
}
```

Which means `BasicStrBuf`
https://github.com/openbmc/stdplus/blob/2f6affd1fdfe52c747a7e0917ab11ccdfa3866a8/include/stdplus/str/buf.hpp#L412C1-L412C34
```
using StrBuf = BasicStrBuf<char>;
```

And this class has only one append method with a different signature that fmtlib requires (see the comment above)
```cpp
template <typename CharT, std::size_t ObjSize = 128,
          typename Allocator = std::allocator<CharT>>
class BasicStrBuf
{
...
    constexpr CharT* append(std::size_t amt)
    {
     ...
    }
...
}
```

Hope that helps.

**Comment by wak-google** [comment]
Oh, they are peering behind the iterator to do their own appending.... Yeah I can fix this.

**Comment by wak-google** [comment]
Fixed by 7aa2a42be328438726a2aff665d649444a1a54d4
